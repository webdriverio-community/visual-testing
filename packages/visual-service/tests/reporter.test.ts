import { describe, it, expect, vi, beforeEach } from 'vitest'
import fs from 'node:fs'
import path from 'node:path'
import { join } from 'node:path'
import VisualReportGenerator from '../src/reporter.js'

vi.mock('fs')
// vi.mock('path')
vi.mock('@wdio/logger', () => import(join(process.cwd(), '__mocks__', '@wdio/logger')))

describe('VisualReportGenerator class', () => {
    const mockDirectoryPath = '/mock-directory'
    let visualReporter: VisualReportGenerator

    beforeEach(() => {
        visualReporter = new VisualReportGenerator({ directoryPath: mockDirectoryPath })
    })

    it('should call readDirectory when calling readJsonFilesRecursively', () => {
        const mockTestData: any[] = [{ '@wdio/visual-service mobile web': { test: 'test1', commandName: 'cmd', instanceData: { platform: { name: 'platform', version: 'v1' } } } }]
        const readDirectorySpy = vi.spyOn(visualReporter as any, 'readDirectory').mockImplementation((dir, testData:any) => {
            testData.push(mockTestData[0])
        })
        // We cast it to any to also test the private method
        const result = (visualReporter as any).readJsonFilesRecursively(mockDirectoryPath)

        expect(readDirectorySpy).toHaveBeenCalledWith(mockDirectoryPath, expect.any(Array))
        expect(result).toEqual(mockTestData)
    })

    it('should read a directory and populate testData array', () => {
        const testData: any[] = []
        const readdirSyncMock = vi.spyOn(fs, 'readdirSync')
            .mockImplementation((dirPath): any => {
                if (dirPath === mockDirectoryPath) {
                    return ['subdir']
                } else if (dirPath === path.join(mockDirectoryPath, 'subdir')) {
                    return ['file-report.json']
                }
                return []
            })
        const statSyncMock = vi.spyOn(fs, 'statSync')
            .mockImplementation((fullPath) => {
                return {
                    isDirectory: () => (fullPath as string).endsWith('subdir'),
                    isFile: () => (fullPath as string).includes('file-report.json'),
                } as fs.Stats
            })
        const readJsonFileSpy = vi.spyOn(visualReporter as any, 'readJsonFile')
            .mockImplementation((filePath, testData: any) => {
                testData.push({ '@wdio/visual-service mobile web': { test: 'test1', commandName: 'cmd', instanceData: { platform: { name: 'platform', version: 'v1' } } } })
            })

        ;(visualReporter as any).readDirectory(mockDirectoryPath, testData)

        expect(readdirSyncMock).toHaveBeenCalledWith(mockDirectoryPath)
        expect(readdirSyncMock).toHaveBeenCalledWith(path.join(mockDirectoryPath, 'subdir'))
        expect(statSyncMock).toHaveBeenCalledWith(path.join(mockDirectoryPath, 'subdir'))
        expect(statSyncMock).toHaveBeenCalledWith(path.join(mockDirectoryPath, 'subdir', 'file-report.json'))
        expect(readJsonFileSpy).toHaveBeenCalledWith(path.join(mockDirectoryPath, 'subdir', 'file-report.json'), testData)
        expect(testData).toHaveLength(1)
    })
})
