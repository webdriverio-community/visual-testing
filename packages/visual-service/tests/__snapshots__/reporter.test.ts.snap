// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VisualReportGenerator class > should group and sort test data correctly 1`] = `
[
  {
    "data": [
      {
        "data": [
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd1",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "browser": {
                "name": "browser1",
                "version": "",
              },
              "platform": {
                "name": "platform1",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test1",
          },
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd1",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "browser": {
                "name": "browser1",
                "version": "",
              },
              "platform": {
                "name": "platform2",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test1",
          },
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd2",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "browser": {},
              "platform": {
                "name": "platform1",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test1",
          },
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd2",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "browser": null,
              "platform": {
                "name": "platform1",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test1",
          },
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd2",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "browser": undefined,
              "platform": {
                "name": "platform1",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test1",
          },
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd2",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "browser": {
                "name": "browser2",
                "version": "",
              },
              "platform": {
                "name": "platform1",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test1",
          },
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd3",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "browser": {
                "name": "browser1",
                "version": "",
              },
              "deviceName": "device1",
              "platform": {
                "name": "platform1",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test1",
          },
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd3",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "browser": {
                "name": "browser1",
                "version": "",
              },
              "deviceName": "device2",
              "platform": {
                "name": "platform1",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test1",
          },
        ],
        "test": "test1",
      },
      {
        "data": [
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd1",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "browser": {
                "name": "browser1",
                "version": "",
              },
              "platform": {
                "name": "platform1",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test2",
          },
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd1",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "browser": {
                "name": "browser1",
                "version": "",
              },
              "platform": {
                "name": "platform3",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test2",
          },
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd2",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "browser": {
                "name": "",
                "version": "",
              },
              "platform": {
                "name": "platform1",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test2",
          },
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd2",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "browser": {
                "name": "browser2",
                "version": "",
              },
              "platform": {
                "name": "platform1",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test2",
          },
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd3",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "platform": {
                "name": "platform1",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test2",
          },
          {
            "boundingBoxes": {
              "diffBoundingBoxes": [],
              "ignoredBoxes": [],
            },
            "commandName": "cmd3",
            "description": "@wdio/visual-service mobile web",
            "fileData": {
              "actualFilePath": "",
              "baselineFilePath": "",
              "diffFilePath": "",
              "fileName": "",
              "size": {
                "actual": {
                  "height": 0,
                  "width": 0,
                },
                "baseline": {
                  "height": 0,
                  "width": 0,
                },
                "diff": {
                  "height": 0,
                  "width": 0,
                },
              },
            },
            "framework": "",
            "instanceData": {
              "browser": {
                "name": "browser3",
                "version": "",
              },
              "platform": {
                "name": "platform2",
                "version": "v1",
              },
            },
            "misMatchPercentage": "0",
            "rawMisMatchPercentage": 0,
            "tag": "",
            "test": "test2",
          },
        ],
        "test": "test2",
      },
    ],
    "description": "@wdio/visual-service mobile web",
  },
]
`;
